### ISSUE TITLE
video-keyboard-response start parameter bug in Safari

### ISSUE THREAD

**Issue by bmchardy** [jspsych/jsPsych#1806]
When setting a `start` parameter for the `video-keyboard-response` plugin in Chrome or Firefox, the video plays from the start time specified (i.e., the plugin works as intended), but in Safari, the stimulus screen appears completely blank with no audio.

**Replication:** Including a `start` time like in the example below appears to work in all browsers _except_ Safari.
```
var stim_video_seg_initial = {
        type: 'video-keyboard-response',
        autoplay: true,
        **start: 10,**
        trial_ends_after_video: true,
        stimulus: [
          'video-location.mp4'
        ],
        choices: jsPsych.NO_KEYS,
};
```

The above code works as intended in Firefox and Chrome. In Safari, the above code only shows a blank screen until the intended video duration has passed.

It looks to me like the issue stems from the `onseeked` listener on line 190 of the `jspsych-video-keyboard-response.js` plugin. It looks to me like this listener is never called on Safari browsers.
```
if(trial.start !== null){
    video_element.pause();
    video_element.currentTime = trial.start;
    video_element.onseeked = function() {
        video_element.style.visibility = "visible";
        if (trial.autoplay) {
            video_element.play();
        }
    }
}
```

This issue seems to replicate across different Mac and iOS devices running Safari. Are you aware of this issue? Are there any fixes that you can think of? Maybe there is an alternative listener to the `onseeked` listener that is specific to Safari devices?

Any input is appreciated. Thanks!

**Comment by becky-gilbert** [comment]
Hi @bmchardy, thanks very much for flagging this, and for tracking down the source of the problem. I quickly checked the [MDN docs for the seeked event](https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement/seeked_event#browser_compatibility), and that says the seeked event _should_ be compatible with Safari, so it's not immediately clear to me why this is happening. Anyway I'll look into it and update here.
Thanks again!

**Comment by becky-gilbert** [comment]
Hi @bmchardy, sorry for the delay getting to this issue. I was able to reproduce it and you're right that the `seeked` event isn't firing in Safari. Apparently in Safari, videos aren't immediately seekable, even when the video is loaded (WHY?! 😧). They do become seekable after they've started playing. So my current hack to get this to work is:
1. Detect Safari \*

Then, if the browser is Safari, 

2. Set up a `playing` event handler that sets the current time to be the trial's start time.
3. Mute and play the video. This will trigger the `playing` event handler, which will change the current time, which will then trigger the `seeked` event handler.
4. The `seeked` event handler is the same as before, except now we also need to (a) un-mute the video, and (b) pause it if it shouldn't auto-play (since we've already started playing it on Safari).

```js
    if(trial.start !== null){
      video_element.pause();
      video_element.onseeked = function() {
        video_element.muted = false;
        video_element.style.visibility = "visible";
        if (trial.autoplay) {
          video_element.play();
        } else {
          video_element.pause();
        }
      }
      var isSafari = /^((?!chrome|android).)*safari/i.test(navigator.userAgent);
      if (!isSafari ) {
        // set current time, which will trigger the seeked event and play video
        video_element.currentTime = trial.start;
      } else {
        // in safari, videos are not immediately seekable, so need to start playing first
        video_element.onplaying = function() {
          video_element.currentTime = trial.start;
        }
        video_element.muted = true;
        video_element.play();
      }
    }
```

\* The Safari detection actually picks up iOS Chrome and probably other iOS browsers. I'm not sure, but I _think_ that that all iOS browsers use Safari under the hood..? In any case, I got the same problem with the video not playing on iOS Chrome, so we need a solution that picks it up. 

Importantly: **I haven't tested this on MacOS.** Would you (or anyone else) be able to test this solution on MacOS please?

I realize that this is hacky and browser detection isn't ideal, but I thought I'd share this with you anyway, in case you've been waiting on a solution. Hopefully we can come up with something better.

**Comment by bmchardy** [comment]
Thanks for all your work on this @becky-gilbert! This looks like a good solution. I can confirm this works on Safari for iPhone and Chrome for iPad. I'll make sure we test this on Mac and will update here with results.

Even though it is a bit hacky to use browser detection, it gets the job done! Your procedure of initializing a `playing` event handler and using this event callback to seek makes a lot of sense. I'm thinking there are a few things we could try such as using this procedure for _all_ browsers instead of just Safari.

When I get a chance, I'll fiddle around with this and see if I can modify your procedure to remove the browser detection component. Thanks again for all your efforts on this!

**Comment by becky-gilbert** [comment]
Thanks for the feedback @bmchardy! Please do let me know if you have any updates and I'll do the same.

**Comment by becky-gilbert** [comment]
Hi @bmchardy! So a little update on this. 

First, there's a bug in the last chunk of code that I posted above. When `autoplay` is `false` and `controls` is `true`, then the code above doesn't work because, when the user hits the play button, the `playing` event will fire and reset the current time, which will then trigger the `seeked` event and pause the video. The solution is to remove both of those event handlers inside the handler functions so that they only fire once.

Second, I switched from Safari detection to detecting any device running iOS or MacOS. So here's what the updated version looks like:
```js
    if(trial.start !== null){
      video_element.pause();
      video_element.onseeked = function() {
        video_element.style.visibility = "visible";
        if (video_element.muted) {
          video_element.muted = false;
        }
        if (trial.autoplay) {
          video_element.play();
        } else if (isiOS) {
          video_element.pause();
        }
        video_element.onseeked = function() {};  // remove seeked event handler
      }
      var isiOS = navigator.userAgent.match(/(iPad|iPhone|iPod|Macintosh)/) !== null;
      if (!isiOS) {
        // set current time, which will trigger the seeked event and play video
        video_element.currentTime = trial.start;
      } else {
        // in MacOS/iOS, videos are not seekable until playing
        video_element.onplaying = function() {
          video_element.currentTime = trial.start;
          video_element.onplaying = function() {};  // remove playing event handler
        }
        video_element.muted = true;
        video_element.play();
      }
    }
```

However, I agree that it might be best to just use this same procedure for all browsers/OSs. Despite the pain it causes me to use this method all the time, it's probably safer to avoid using user agent strings and browser/OS detection. Plus it makes the code a bit shorter.
```js
    if(trial.start !== null){
      video_element.pause();
      video_element.onseeked = function() {
        video_element.style.visibility = "visible";
        video_element.muted = false;
        if (trial.autoplay) {
          video_element.play();
        } else {
          video_element.pause();
        }
        video_element.onseeked = function() {};
      }
      video_element.onplaying = function() {
        video_element.currentTime = trial.start;
        video_element.onplaying = function() {};
      }
      video_element.muted = true;
      video_element.play();
    }
```
I'm thinking we should go with the latter solution, so let me know if you have a chance to try this and if it works for you. And it'd still be great if you could test this on MacOS, if possible. It'd be great if we could include this fix in the next release (7.0), which we're hoping will be ready in the next few weeks.

**Comment by bmchardy** [comment]
Hi @becky-gilbert! Thanks for your update on this. I agree that the latter solution is probably best because of its simplicity and because it no longer requires using browser detection methods. I've had the chance to test this latter solution on iOS (Safari), iPadOS (both Safari and Chrome), and MacOS (Safari). No issues to report and all behaviour seems as intended.

I believe this is a strong fix! Thanks for all of your hard work on this!

**Comment by becky-gilbert** [comment]
@bmchardy great, thanks very much for the feedback! This fix will be added to the next release. Thanks for reporting the issue, and for your patience 😃

**Comment by bmchardy** [comment]
@becky-gilbert great! Thank you!

**Comment by jodeleeuw** [comment]
@becky-gilbert Will you patch this in now that `modularization` is ready for small updates?

**Comment by becky-gilbert** [comment]
@jodeleeuw this is now fixed on the `video-start-patch` branch, which is based off `modularization`.  So we could either merge into `modularization` now, or wait and then merge into `main`.

I haven't added a changeset file yet because I wasn't sure if it makes sense to flag this as a patch for the `video-*` plugins and thus bump their version numbers, given that these plugins haven't been published as npm packages yet?

**Comment by bjoluc** [comment]
> I haven't added a changeset file yet because I wasn't sure if it makes sense to flag this as a patch for the video-* plugins and thus bump their version numbers, given that these plugins haven't been published as npm packages yet?

Feel free to do so. The version will be [bumped from 0.1.0 to 1.0.0 anyway](https://github.com/jspsych/jsPsych/blob/9d8cff0121f458470b1bfe6b92a58d26fb743eb6/.changeset/major.md), so an additional patch change won't affect the version number, but it will be added to the release notes and changelog.
